File: src/scala/scaled/project/JDKProject.scala
//
// Scaled Java Project Support - helps Scaled Project framework grok Java.
// http://github.com/scaled/java-project/blob/master/LICENSE

package scaled.project

import codex.extract.SourceSet
import java.nio.file.Path
import java.util.function.Predicate
import java.util.zip.ZipEntry
import scaled._
import scaled.pacman.JDK

/** A special project for use when another project depends on the JDK. This will eventually handle
  * locating JDKs on different platforms and turning the contents of src.zip into a Codex, and
  * whatever other special casery arises.
  */
class JDKProject (ps :ProjectSpace, val jdk :JDK)
    extends AbstractZipFileProject(ps, Project.Root(jdk.root)) {
  import Project._

  override val zipPaths = Seq(jdk.root)

  override protected def computeMeta (oldMeta :Meta) = {
    Future.success(oldMeta.copy(
      name = s"jdk-${jdk.version}",
      ids = Seq(PlatformId(JavaPlatform, jdk.majorVersion))
    ))
  }

  override def summarizeSources = Map(
    "java" -> new SourceSet.Archive(zipPaths.head, new Predicate[ZipEntry] {
      def test (entry :ZipEntry) = entry.getName.startsWith("java")
    })
  )
}

object JDKProject {
  import Project._

  @Plugin(tag="project-finder")
  class FinderPlugin extends ProjectFinderPlugin("jdk", true, classOf[JDKProject]) {
    override def checkRoot (root :Path) :Int = if (JDK.jdks.exists(_.root == root)) 1 else -1

    override def apply (id :Id) = id match {
      case PlatformId(JavaPlatform, version) =>
        JDK.jdks.find(_.majorVersion == version).map(jdk => seed(Root(jdk.root), jdk :: Nil))
      case _ => None
    }

    override protected def injectArgs (root :Root) = List(JDK.jdks.find(_.root == root.path) || {
      throw new IllegalArgumentException(s"Invalid JDK root: $root")
    })
  }
}

File: src/java/scaled/project/JDKProject.java
//
// Scaled Java Mode - support for editing Java code
// https://github.com/scaled/java-mode/blob/master/LICENSE

package scaled.project;

import codex.extract.SourceSet;
import java.nio.file.Path;
import java.util.function.Predicate;
import java.util.zip.ZipEntry;
import scaled.*;
import scaled.pacman.JDK;

/** A special project for use when another project depends on the JDK. This will eventually handle
  * locating JDKs on different platforms and turning the contents of src.zip into a Codex, and
  * whatever other special casery arises.
  */
public class JDKProject extends Project {

  @Plugin(tag="project-finder")
  public static class FinderPlugin extends ProjectFinderPlugin {
    public FinderPlugin () {
      super("jdk", true, JDKProject.class);
    }

    @Override public int checkRoot (Path root) {
      return (Seq.view(JDK.jdks()).exists(jdk -> jdk.root().equals(root))) ? 1 : -1;
    }

    @Override public Option<Project.Seed> apply (Id id) {
      if (id instanceof PlatformId) {
        String platform = ((PlatformId)id).platform();
        String version = ((PlatformId)id).version();
        if (platform.equals(Project.JavaPlatform())) {
          return Seq.view(JDK.jdks()).find(jdk -> jdk.majorVersion().equals(version)).
            map(jdk -> seed(new Project.Root(jdk.root(), ""), List.apply(jdk)));
        }
      }
      return Option.none();
    }

    @Override public List<Object> injectArgs (Project.Root root) {
      JDK jdk = Seq.view(JDK.jdks()).find(j -> j.root().equals(root.path())).getOrElse(() -> {
        throw new IllegalArgumentException("Invalid JDK root: " + root);
      });
      return Std.list(jdk);
    }
  }

  public final JDK jdk;

  public JDKProject (ProjectSpace ps, JDK jdk) {
    super(ps, new Project.Root(jdk.root(), ""));
    this.jdk = jdk;
  }

  @Override public Future<Meta> computeMeta (Meta oldMeta) {
    // add a filer component for our zip file(s)
    addComponent(Filer.class, new ZipFiler(Std.seq(jdk.root())));

    // add a sources component that groks our zip-based source files
    addComponent(Sources.class, new Sources(Std.seq(jdk.root())) {
      @Override public Map<String, SourceSet> summarize () {
        return Map.<String, SourceSet>builder().put(
          "java", new SourceSet.Archive(jdk.root(), entry -> entry.getName().startsWith("java"))
        ).build();
      }
    });

    return Future.success(new Meta(
      "jdk-" + jdk.version(),
      Std.seq(new PlatformId(Project.JavaPlatform(), jdk.majorVersion()))
    ));
  }
}
